File: src/main/java/club/sk1er/mods/core/universal/wrappers/UniversalPlayer.java
package club.sk1er.mods.core.universal.wrappers;

import club.sk1er.mods.core.universal.UniversalMinecraft;
import club.sk1er.mods.core.universal.wrappers.message.UniversalTextComponent;

import java.util.UUID;

//#if MC>=11502
//$$ import net.minecraft.client.entity.player.ClientPlayerEntity;
//#else
import net.minecraft.client.entity.EntityPlayerSP;
//#endif

public class UniversalPlayer {
    public static boolean hasPlayer() {
        return getPlayer() != null;
    }

    //#if MC>=11502
    //$$ public static ClientPlayerEntity getPlayer() {
    //#else
    public static EntityPlayerSP getPlayer() {
    //#endif
        //#if MC>=11202
        //$$ return UniversalMinecraft.getMinecraft().player;
        //#else
        return UniversalMinecraft.getMinecraft().thePlayer;
        //#endif
    }

    public static void sendClientSideMessage(UniversalTextComponent chatMessage) {
        //#if MC>=11602
        //$$ getPlayer().sendMessage(chatMessage, null);
        //#else
        getPlayer().addChatMessage(chatMessage);
        //#endif
    }

    public static UUID getUUID() {
        return UniversalMinecraft.getMinecraft().getSession().getProfile().getId();
    }

    public static double getPosX() {
        if (!hasPlayer())
            throw new NullPointerException("UniversalPlayer.getPosX() called with no existing Player");

        //#if MC>=11502
        //$$ return getPlayer().getPosX();
        //#else
        return getPlayer().posX;
        //#endif
    }

    public static double getPosY() {
        if (!hasPlayer())
            throw new NullPointerException("UniversalPlayer.getPosY() called with no existing Player");

        //#if MC>=11502
        //$$ return getPlayer().getPosY();
        //#else
        return getPlayer().posY;
        //#endif
    }

    public static double getPosZ() {
        if (!hasPlayer())
            throw new NullPointerException("UniversalPlayer.getPosZ() called with no existing Player");

        //#if MC>=11502
        //$$ return getPlayer().getPosZ();
        //#else
        return getPlayer().posZ;
        //#endif
    }

    public static double getPrevPosX() {
        if (!hasPlayer())
            throw new NullPointerException("UniversalPlayer.getPrevPosX() called with no existing Player");

        return getPlayer().prevPosX;
    }

    public static double getPrevPosY() {
        if (!hasPlayer())
            throw new NullPointerException("UniversalPlayer.getPrevPosY() called with no existing Player");

        return getPlayer().prevPosY;
    }

    public static double getPrevPosZ() {
        if (!hasPlayer())
            throw new NullPointerException("UniversalPlayer.getPrevPosZ() called with no existing Player");

        return getPlayer().prevPosZ;
    }
}

File: src/main/kotlin/club/sk1er/mods/core/universal/wrappers/UniversalPlayer.kt
package club.sk1er.mods.core.universal.wrappers

import club.sk1er.mods.core.universal.UniversalMinecraft
import club.sk1er.mods.core.universal.utils.MCEntityPlayerSP
import club.sk1er.mods.core.universal.wrappers.message.UniversalTextComponent
import java.util.*

object UniversalPlayer {
    @JvmStatic
    fun getPlayer(): MCEntityPlayerSP? {
        //#if MC>=11202
        //$$ return UniversalMinecraft.getMinecraft().player
        //#else
        return UniversalMinecraft.getMinecraft().thePlayer
        //#endif
    }

    @JvmStatic
    fun hasPlayer() = getPlayer() != null

    @JvmStatic
    fun sendClientSideMessage(message: UniversalTextComponent) {
        //#if FABRIC
        //$$ getPlayer()!!.sendMessage(message, false)
        //#elseif MC>=11602
        //$$ getPlayer()!!.sendMessage(message, null)
        //#elseif MC>=11202
        //$$ getPlayer()!!.sendChatMessage(message.formattedText)
        //#else
        getPlayer()!!.addChatMessage(message)
        //#endif
    }

    @JvmStatic
    fun getUUID(): UUID {
        return UniversalMinecraft.getMinecraft().session.profile.id
    }

    @JvmStatic
    fun getPosX(): Double {
        //#if FABRIC
        //$$ return getPlayer()?.pos?.x
        //#else
        return getPlayer()?.posX
        //#endif
            ?: throw NullPointerException("UniversalPlayer.getPosX() called with no existing Player")
    }

    @JvmStatic
    fun getPosY(): Double {
        //#if FABRIC
        //$$ return getPlayer()?.pos?.y
        //#else
        return getPlayer()?.posY
        //#endif
            ?: throw NullPointerException("UniversalPlayer.getPosY() called with no existing Player")
    }

    @JvmStatic
    fun getPosZ(): Double {
        //#if FABRIC
        //$$ return getPlayer()?.pos?.z
        //#else
        return getPlayer()?.posZ
        //#endif
            ?: throw NullPointerException("UniversalPlayer.getPosZ() called with no existing Player")
    }

    @JvmStatic
    fun getPrevPosX(): Double {
        //#if FABRIC
        //$$ return getPlayer()?.prevX
        //#else
        return getPlayer()?.prevPosX
        //#endif
            ?: throw NullPointerException("UniversalPlayer.getPrevPosX() called with no existing Player")
    }

    @JvmStatic
    fun getPrevPosY(): Double {
        //#if FABRIC
        //$$ return getPlayer()?.prevY
        //#else
        return getPlayer()?.prevPosY
        //#endif
            ?: throw NullPointerException("UniversalPlayer.getPrevPosY() called with no existing Player")
    }

    @JvmStatic
    fun getPrevPosZ(): Double {
        //#if FABRIC
        //$$ return getPlayer()?.prevZ
        //#else
        return getPlayer()?.prevPosZ
        //#endif
            ?: throw NullPointerException("UniversalPlayer.getPrevPosZ() called with no existing Player")
    }
}
